llvm/lib/MCCAS/MCCASObjectV1.cpp

```diff
@@ -2014,6 +2011,106 @@ MCCASBuilder::createDebugStringRefs() {
+/// is the index into the \p RelocationBuffer and is used to access the current
+/// relocation that has to be partitioned, we need to do this because sometimes
+/// we only have access to all the relocations for a section and a relocation
+/// may, in that case, not belong to the current fragment.
+
+/// FragmentIndex: It denotes the index into the FragmentContents that is used
+/// to copy the data that deduplicates in the \p FinalFragmentContents.
+
+/// PrevOffset: A relocation can sometimes be divided into multiple parts, all
+/// of them share the same offset, but describe only one Addend, if we already
+/// copied the addend out of the FragmentContents at a particular offset, we
+/// should skip all relocations that matches the same offset.
+
+/// RelocOffsetInFragment: This is used to denote the offset of the relocation
+/// in the current Fragment. Relocation offsets are always from the start of the
+/// section, we need to normalize this value to start at the start of the
+/// Fragment instead.
+static void
+partitionFragment(const MCAsmLayout &Layout, SmallVector<char, 0> &Addends,
+                  SmallVector<char, 0> &FinalFragmentContents,
+                  ArrayRef<MachO::any_relocation_info> RelocationBuffer,
+                  const MCFragment &Fragment, uint64_t &RelocationBufferIndex,
+                  bool IsLittleEndian) {
+  auto FragmentContents = getFragmentContents(Fragment);
+  // Index in current Fragment.
+  uint64_t FragmentIndex = 0;
+  int64_t PrevOffset = -1;
+  // Relocations are stored in the Section.
+  for (; RelocationBufferIndex < RelocationBuffer.size();
+       RelocationBufferIndex++) {
+    auto Reloc = RelocationBuffer[RelocationBufferIndex];
+    uint32_t RelocOffset = getRelocationOffset(Reloc);
+    if (PrevOffset == RelocOffset)
+      continue;
+    uint64_t FragmentOffset = Layout.getFragmentOffset(&Fragment);
+    if (RelocOffset < FragmentOffset + FragmentContents.size()) {
+      uint32_t RelocOffsetInFragment = RelocOffset - FragmentOffset;
+      // Copy contents of fragment upto relocation addend data in the
+      // FinalFragmentContents.
+      FinalFragmentContents.append(FragmentContents.data() + FragmentIndex,
+                                   FragmentContents.data() +
+                                       RelocOffsetInFragment);
+      // Addend belongs to the current fragment, copy the addend into the
+      // Addend vector.
+      Addends.append(FragmentContents.data() + RelocOffsetInFragment,
+                     FragmentContents.data() + RelocOffsetInFragment +
+                         getRelocationSize(Reloc, IsLittleEndian));
+      FragmentIndex =
+          RelocOffsetInFragment + getRelocationSize(Reloc, IsLittleEndian);
+      PrevOffset = RelocOffset;
+    } else
+      break;
+  }
+  // Copy any leftover fragment contents into the FinalFragmentContents.
+  if (FragmentContents.size() > FragmentIndex)
```

Comment:
The combination of this condition and the append call looks odd to me. Aren't we appending an empty set when the condition triggers?

Reply:
Not always, if there are no relocations for this fragment then we are appending all the fragment contents, if there are contents in the fragment after the last relocation addend has been copied, then we need to copy the remaining fragment contents into the `FinalFragmentContents`